File: Solutions/Vertical_Order_Traversal_of_a_Binary_Tree/Vertical_Order_Traversal_of_a_Binary_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalTraversal = function(root) {
    if (root == null) {
        return [];
    }
    
    const queue = [];
    
    queue.push([root, 0, 0]);
    const list = [];
    while(queue.length > 0) {
        const [node, row, column] = queue.shift();
        
        
        if (node) {
            queue.push([node.left, row + 1, column - 1])
            queue.push([node.right, row + 1, column + 1])
            list.push([node.val, row, column]);
        }
    }
  
  const compare = (a, b) => {
    if (a[2] - b[2] === 0 ) {
      
      if (a[1] - b[1] === 0) {
        return a[0] - b[0];
      }
    }
    
    return a[2] - b[2]
  }
  
   list.sort(compare)
  
  const map = new Map();
  for(let i = 0; i < list.length; i++) {
    const [value, row, column] = list[i]
    
    if (map.has(column)) {
      map.get(column).push(value)
    } else {
      map.set(column, [value]);
    }
  }
  
  return [...map.values()]
};



//Most common:
var verticalTraversal = function(root) {
    let min = 0;
    let max = 0;
    const map = new Map();
    
    (function buildMap(root, row, col) {
        if(!root) return;
        map.set(col, (map.get(col) || []).concat({row, val: root.val}));
        min = Math.min(min,col);
        max = Math.max(max, col);
        buildMap(root.left, row+1, col-1, map);
        buildMap(root.right, row+1, col+1, map);
    })(root, 0, 0)
    
    let result = [];
    for (let i = min; i <= max; i++) {
        let nodes = map.get(i);
        nodes.sort((a,b)=> {
            if(a.row !== b.row) return a.row-b.row;
            else return a.val-b.val;
        });
        let vals = nodes.map(x=>x.val);
        result.push(vals);
    }
    
    return result;
};

//Fastest:
var verticalTraversal = function(root) {
    function node(row, val) {
        return {row, val};
    }
    let min = 0;
    let max = 0;
    let map = build(root, 0,0,new Map());
    let result = [];
    for(let i=min; i<=max; i++) {
        let nodes = map.get(i);
        nodes.sort((a,b)=> {
            if(a.row !== b.row) return a.row-b.row;
            else return a.val-b.val;
        });
        let vals = nodes.map(x=>x.val);
        result.push(vals);
    }
    return result;
    
    function build(root, row, col, map) {
        if(!root) return map;
        if(!map.has(col)) map.set(col, []);
        map.get(col).push(new node(row, root.val));
        min = Math.min(min,col);
        max = Math.max(max, col);
        map = build(root.left, row+1, col-1, map);
        return build(root.right, row+1, col+1, map);
    }
};

File: Solutions/Vertical_Order_Traversal_of_a_Binary_Tree/Vertical_Order_Traversal_of_a_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import defaultdict,deque
class Solution:
    def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        mp = defaultdict(lambda:[])
        minh = 0
        maxh = 0
        q = deque()
        q.append( (root,0) )
        h= 0 
        while q:
            l = len(q)
            h-=1
            tmp = defaultdict(lambda :[])
            for _ in range(l):
                r,x = q.popleft()
                tmp[x].append(r.val)
                if r.left:
                    q.append( (r.left,x-1) )
                if r.right:
                    q.append( (r.right,x+1) )
            
            for k in tmp:
                tmp[k].sort()
                mp[k].extend(tmp[k])
                
        ans = [ (k,v) for k,v in mp.items() ]
        ans.sort(key = lambda x:x[0] )
        ans = [ x[1] for x in ans ]
        return ans

#Fastest
class Solution:
     def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        
        if not root:
            return []
        
        cols = collections.defaultdict(list)
        queue = deque([(root, 0, 0)])
        min_col = max_col = 0
        result = []
        
        while queue:
            
            node, row, col = queue.popleft()
            if node: 
                cols[col].append((row, node.val))
                min_col = min(min_col, col)
                max_col = max(max_col, col)
                
                if node.left:
                    queue.append((node.left, row + 1, col - 1))
                
                if node.right:
                    queue.append((node.right, row + 1, col + 1))
                    
        
        for col in range(min_col, max_col + 1):
            result.append([val for row, val in sorted(cols[col])])
            
        return result